pkg/network/ebpf/c/protocols/protocol-classification-helpers.h

```diff
@@ -8,33 +8,55 @@
 long bpf_skb_load_bytes_with_telemetry(const void *skb, u32 offset, void *to, u32 len) {return 0;}
 #endif
 
 #define CHECK_PRELIMINARY_BUFFER_CONDITIONS(buf, buf_size, min_buff_size)   \
-        if (buf_size < min_buff_size) {                                     \
-            return false;                                                   \
-        }                                                                   \
+    if (buf_size < min_buff_size) {                                         \
+        return false;                                                       \
+    }                                                                       \
                                                                             \
-        if (buf == NULL) {                                                  \
-            return false;                                                   \
-        }                                                                   \
+    if (buf == NULL) {                                                      \
+        return false;                                                       \
+    }                                                                       \
 
 // The method checks if the given buffer starts with the HTTP2 marker as defined in https://datatracker.ietf.org/doc/html/rfc7540.
 // We check that the given buffer is not empty and its size is at least 24 bytes.
-static __always_inline bool is_http2(const char* buf, __u32 buf_size) {
+static __always_inline bool is_http2_preface(const char* buf, __u32 buf_size) {
     CHECK_PRELIMINARY_BUFFER_CONDITIONS(buf, buf_size, HTTP2_MARKER_SIZE)
 
-#define HTTP2_SIGNATURE "PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n"
+#define HTTP2_PREFACE "PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n"
 
-    bool match = !bpf_memcmp(buf, HTTP2_SIGNATURE, sizeof(HTTP2_SIGNATURE)-1);
+    bool match = !bpf_memcmp(buf, HTTP2_PREFACE, sizeof(HTTP2_PREFACE)-1);
 
     return match;
 }
 
+// According to the https://www.rfc-editor.org/rfc/rfc7540#section-3.5
+// an HTTP2 server must reply with a settings frame to the preface of HTTP2.
+// The settings frame must not be related to the connection (stream_id == 0) and the length should be a multiplication
+// of 6 bytes.
+static __always_inline bool is_http2_server_settings(const char* buf, __u32 buf_size) {
+    CHECK_PRELIMINARY_BUFFER_CONDITIONS(buf, buf_size, HTTP2_FRAME_HEADER_SIZE)
+
+    struct http2_frame frame_header;
+    if (!read_http2_frame_header(buf, buf_size, &frame_header)) {
+        return false;
+    }
+
+    return frame_header.type == kSettingsFrame && frame_header.stream_id == 0 && frame_header.length % HTTP2_SETTINGS_SIZE == 0;
+}
+
+// The method checks if the given buffer starts with the HTTP2 marker as defined in https://datatracker.ietf.org/doc/html/rfc7540.
+// We check that the given buffer is not empty and its size is at least 24 bytes.
+static __always_inline bool is_http2(const char* buf, __u32 buf_size) {
```

Comment:
Is this getting called from somewhere?

Reply:
yes, https://github.com/DataDog/datadog-agent/blob/guy.arbitman/USMO-146-improve-http2-classification/pkg/network/ebpf/c/protocols/protocol-classification-helpers.h#L100

Reply:
namely `classify_protocol` function